python_modules/dagster/dagster/_core/execution/run_metrics_thread.py

```diff
@@ -0,0 +1,252 @@
+
+            if python_metrics_enabled:
+                python_metrics = _get_python_runtime_metrics()
+                metrics.update(python_metrics)
+
+            if len(metrics) > 0:
+                _report_run_metrics(
+                    instance,
+                    dagster_run=dagster_run,
+                    metrics=metrics,
+                    run_tags=run_tags,
+                )
+
+        except:
+            logging.error(
+                "Exception during capture of metrics, will cease capturing", exc_info=True
+            )
+            return False  # terminate the thread safely without interrupting the main thread
+
+        sleep(polling_interval)
+    if logger:
+        logging.debug("Shutting down metrics capture thread")
+    return True
+
+
+def start_run_metrics_thread(
+    instance: DagsterInstance,
+    dagster_run: DagsterRun,
+    container_metrics_enabled: Optional[bool] = True,
+    python_metrics_enabled: Optional[bool] = False,
+    logger: Optional[logging.Logger] = None,
+    polling_interval: float = DEFAULT_RUN_METRICS_POLL_INTERVAL_SECONDS,
+) -> Tuple[threading.Thread, threading.Event]:
+    check.inst_param(instance, "instance", DagsterInstance)
+    check.inst_param(dagster_run, "dagster_run", DagsterRun)
+    check.opt_inst_param(logger, "logger", logging.Logger)
+    check.opt_bool_param(container_metrics_enabled, "container_metrics_enabled")
+    check.opt_bool_param(python_metrics_enabled, "python_metrics_enabled")
+    check.float_param(polling_interval, "polling_interval")
+
+    container_metrics_enabled = container_metrics_enabled and _process_is_containerized()
+
+    # TODO - ensure at least one metrics source is enabled
+    assert container_metrics_enabled or python_metrics_enabled, "No metrics enabled"
+
+    if logger:
+        logger.debug("Starting run metrics thread")
+
+    instance.report_engine_event(
+        f"Starting run metrics thread with container_metrics_enabled={container_metrics_enabled} and "
+        f"python_metrics_enabled={python_metrics_enabled}",
+        dagster_run=dagster_run,
+        run_id=dagster_run.run_id,
+    )
+
+    shutdown_event = threading.Event()
+    thread = threading.Thread(
+        target=_capture_metrics,
+        args=(
+            instance,
+            dagster_run,
+            container_metrics_enabled,
+            python_metrics_enabled,
+            shutdown_event,
+            polling_interval,
+            logger,
+        ),
+        name="run-metrics",
+    )
+    thread.start()
+    return thread, shutdown_event
+
+
+def stop_run_metrics_thread(
+    thread: threading.Thread,
+    stop_event: threading.Event,
+    timeout: Optional[int] = DEFAULT_RUN_METRICS_SHUTDOWN_SECONDS,
+) -> bool:
+    thread = check.not_none(thread)
+    stop_event = check.not_none(stop_event)
+
+    stop_event.set()
+    if thread.is_alive():
+        thread.join(timeout=timeout)
+
+    return not thread.is_alive()
```

Comment:
will this ever not be true at this point? If the timeout in the join fails it will raise an exception - doesn't seem like this method needs to return anything and instead will raise an exception if it fails

Reply:
I am not sure about that:

https://docs.python.org/3/library/threading.html#threading.Thread.join

> As [join()](https://docs.python.org/2/library/threading.html#threading.Thread.join) always returns None, you must call [isAlive()](https://docs.python.org/2/library/threading.html#threading.Thread.isAlive) after [join()](https://docs.python.org/2/library/threading.html#threading.Thread.join) to decide whether a timeout happened – if the thread is still alive, the [join()](https://docs.python.org/2/library/threading.html#threading.Thread.join) call timed out.

The only explicit case where there is an exception thrown is if the join would cause a deadlock.